File: core/src/com/gdxjam/magellan/ships/PlayerShip.java
package com.gdxjam.magellan.ships;

import aurelienribon.tweenengine.Timeline;
import aurelienribon.tweenengine.Tween;
import aurelienribon.tweenengine.TweenEquations;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ObjectMap;
import com.gdxjam.magellan.*;
import com.gdxjam.magellan.drones.Drone;
import com.gdxjam.magellan.drones.DroneRoutineScouting;
import com.gdxjam.magellan.screen.BaseScreen;

/**
 * Created by lolcorner on 20.12.2015.
 */
public class PlayerShip extends Ship implements IInteractable {

    public int HUMANS = 10000;
    public Array<Integer> drones = new Array();
    public Array<ShopItem> inventory;

    public PlayerShip(Sector sector) {
        super(sector);
        faction = Factions.PLAYER;
        inventory = new Array<ShopItem>();
        setSectorsDiscovered();
        drones.add(4);
    }

    public void moveTo(Sector sector) {
        super.moveTo(sector);
        setSectorsDiscovered();
        tweenManager.killAll();
        Timeline.createSequence()
                .push(Tween.to(this.spriteVessel, SpriteAccessor.ROTATION, 0.3f).target((float)Math.atan2(sector.position.y - lastSector.position.y, sector.position.x - lastSector.position.x)*180f/(float)Math.PI-90f))
                .push(Tween.to(this.spriteVessel, SpriteAccessor.POSITION_XY, 0.5f).target(sector.position.x + 20, sector.position.y - 30).ease(TweenEquations.easeInOutQuint))
                .push(Tween.to(this.spriteVessel, SpriteAccessor.ROTATION, 1f).target(50).ease(TweenEquations.easeInOutCubic))
                .push(Tween.to(this.spriteVessel, SpriteAccessor.POSITION_XY, 0.5f).target(sector.position.x + 12, sector.position.y - 22).ease(TweenEquations.easeInOutCubic)).delay(-0.2f)
        .start(tweenManager);
    }

    public void releaseDrone(int level){
        drones.removeValue(level, false);
        Drone drone = new Drone(this.sector, level);
        drone.faction = faction;
        drone.addRoutine(new DroneRoutineScouting(drone));
        MagellanGame.gameState.updateNumberOfDrones();
    }

    private void setSectorsDiscovered() {
        sector.visited = true;
        sector.discovered = true;
        for(Sector _sector : sector.connectedSectors){
            _sector.discovered = true;
        }
    }

    @Override
    public void prepareRenderingOnMap() {
        //spriteDot = new Sprite(MagellanGame.assets.get("circle.png", Texture.class));
        //spriteDot.setSize(24,24);
        //spriteDot.setColor(Color.YELLOW);

        spriteVessel = new Sprite(MagellanGame.assets.get("map_playership.png", Texture.class));
        spriteVessel.setSize(20, 20);
        spriteVessel.setOriginCenter();
        spriteVessel.setPosition(sector.position.x + 12, sector.position.y - 22);
        spriteVessel.setRotation(50);
        spriteVessel.setColor(MagellanColors.FACTION_PLAYER);
    }

    @Override
    public void renderOnMap(SpriteBatch batch, float delta) {
        super.render(delta);
        //spriteDot.setPosition(sector.position.x - spriteDot.getWidth()/2, sector.position.y - spriteDot.getHeight()/2);
        //spriteDot.draw(batch);

        spriteVessel.draw(batch);
    }

    @Override
    public ObjectMap<String, Interaction> getInteractions(GameObj with) {
        ObjectMap<String, Interaction> interactions = new ObjectMap();
        if (submenuOpen == "") {
            if (drones.size > 0) {
                interactions.put("release drone", new Interaction() {
                    @Override
                    public void interact() {
                        submenuOpen = "releasedrone";
                        showInteractionWindow();
                    }
                });
            }
        }
        if (submenuOpen == "releasedrone") {
            if (drones.contains(1,true)) {
                interactions.put("Level 1", new Interaction() {
                    @Override
                    public void interact() {
                        releaseDrone(1);
                        closeWindow();
                    }
                });
            }
            if (drones.contains(2,true)) {
                interactions.put("Level 2", new Interaction() {
                    @Override
                    public void interact() {
                        releaseDrone(2);
                        closeWindow();
                    }
                });
            }
            if (drones.contains(3,true)) {
                interactions.put("Level 3", new Interaction() {
                    @Override
                    public void interact() {
                        releaseDrone(3);
                        closeWindow();
                    }
                });
            }
            if (drones.contains(4,true)) {
                interactions.put("Level 4", new Interaction() {
                    @Override
                    public void interact() {
                        releaseDrone(4);
                        closeWindow();
                    }
                });
            }
            if (drones.contains(5,true)) {
                interactions.put("Level 5", new Interaction() {
                    @Override
                    public void interact() {
                        releaseDrone(5);
                        closeWindow();
                    }
                });
            }
        }

        return interactions;
    }

    @Override
    public String getTitle() {
        return "The Trinidad";
    }

    @Override
    public String getInfo() {
        String s = "Your ship.";
        s += "\nHealth: " + getHealth();
        s += "\nFrozen Humans: " + HUMANS;
        s += "\nDrones: " + drones.toString(", ");
        s += "\nEquipment: " + inventory.toString(", ");
        return s;
    }
}

File: core/src/com/gdxjam/magellan/ships/PlayerShip.kt
package com.gdxjam.magellan.ships

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.ParticleEffect
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Stack
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.OrderedMap
import com.gdxjam.magellan.MagellanGame
import com.gdxjam.magellan.drones.Drone
import com.gdxjam.magellan.gameobj.GameObj
import com.gdxjam.magellan.gameobj.IDestroyable
import com.gdxjam.magellan.gameobj.IInteractable
import com.gdxjam.magellan.models.Sector
import com.gdxjam.magellan.shopitem.ShopItem
import com.gdxjam.magellan.utils.textureSprite

/**
 * Created by lolcorner on 20.12.2015.
 */
class PlayerShip(sector: Sector) : Ship(sector), IInteractable {

    var maxHealth: Int = 0
    var HUMANS = 10000
    var drones: Array<Int> = Array()
    var inventory: Array<ShopItem>
    var mineResourcesPerTick = 10

    private lateinit var trail: ParticleEffect

    override val actor: Actor
        get() = Stack().apply {
            addActor(Image(MagellanGame.assets.get("sectorview_ship.png", Texture::class.java)))
            addActor(Image(MagellanGame.assets.get("sectorview_ship_shield.png", Texture::class.java)))
            children.get(1).setColor(1f, 1f, 1f, 0f)
        }

    override val title = "The Trinidad"

    override val info: String
        get() = """
            Your ship.
            Health: $health/$maxHealth
            Attack: $attack
            Shield: ${Math.round(shield * 100)}%
            Frozen Humans: $HUMANS
            Drones: ${drones.joinToString(prefix = "", postfix = "")}
            Equipment: ${inventory.joinToString(prefix = "", postfix = "")}""".trimIndent()

    init {
        faction = GameObj.Factions.PLAYER
        health = 15
        maxHealth = health
        attack = 4
        shield = 0.15f
        inventory = Array()

        setSectorsDiscovered()

        drones.add(1)
    }

    override fun shootAt(target: IDestroyable): Int {
        MagellanGame.gameState.aiHostility = 5

        return super.shootAt(target)
    }

    override fun moveTo(sector: Sector) {
        super.moveTo(sector)

        setSectorsDiscovered()
        MagellanGame.soundFx.shipJump.play(0.3f)

        val particlePosition = lastParkingPosition.cpy().lerp(parkingPosition, 0.10f)

        trail.emitters.forEach { em ->
            em.setPosition(particlePosition.x + spriteVessel.width / 2, particlePosition.y + spriteVessel.height / 2)
            em.angle.setHigh(flightAngle - 90f - 10f, flightAngle - 90 + 10)
            em.angle.setLow(flightAngle - 90)
        }

        trail.start()
    }

    override fun destroy() {
        MagellanGame.soundFx.doomed.play()

        dispose()
    }

    fun releaseDrone(level: Int) {
        drones.removeValue(level, false)

        val drone = Drone(this.sector, level)
        drone.faction = faction

        MagellanGame.gameState.updateNumberOfDrones()
    }

    private fun setSectorsDiscovered() {
        sector.visited = true
        sector.discovered = true

        sector.connectedSectors.forEach { _sector ->
            _sector.connectedSectors.forEach { __sector -> __sector.discovered = true }
            _sector.discovered = true
            _sector.visited = true
        }
    }

    override fun prepareRenderingOnMap() {
        super.prepareRenderingOnMap()

        spriteVessel = MagellanGame.assets.textureSprite("map_playership.png")
        spriteVessel.setSize(20f, 20f)
        spriteVessel.setOriginCenter()

        sectorSlot = 1
        getParkingPosition()

        spriteVessel.setPosition(parkingPosition.x, parkingPosition.y)
        spriteVessel.rotation = parkingAngle

        trail = ParticleEffect()
        trail.load(Gdx.files.internal("ship_trail.p"), Gdx.files.internal(""))
        trail.setPosition(parkingPosition.x, parkingPosition.y)
        trail.scaleEffect(0.3f)
    }

    override fun renderOnMap(batch: SpriteBatch, delta: Float) {
        super.render(delta)

        spriteVessel.draw(batch)

        trail.draw(batch, delta)
        //if (trail.isComplete()) trail.reset();
    }

    override fun getInteractions(with: GameObj): OrderedMap<String, IInteractable.Interaction> {
        val interactions = OrderedMap<String, IInteractable.Interaction>()

        if (submenuOpen == "") {
            if (drones.size > 0) {
                interactions.put("release drone", object : IInteractable.Interaction {
                    override fun interact() {
                        submenuOpen = "releasedrone"
                        showInteractionWindow()
                    }
                })
            }
        }
        if (submenuOpen == "releasedrone") {
            if (drones.contains(1, true)) {
                interactions.put("Level 1", object : IInteractable.Interaction {
                    override fun interact() {
                        releaseDrone(1)
                        closeWindow()
                    }
                })
            }
            if (drones.contains(2, true)) {
                interactions.put("Level 2", object : IInteractable.Interaction {
                    override fun interact() {
                        releaseDrone(2)
                        closeWindow()
                    }
                })
            }
            if (drones.contains(3, true)) {
                interactions.put("Level 3", object : IInteractable.Interaction {
                    override fun interact() {
                        releaseDrone(3)
                        closeWindow()
                    }
                })
            }
            if (drones.contains(4, true)) {
                interactions.put("Level 4", object : IInteractable.Interaction {
                    override fun interact() {
                        releaseDrone(4)
                        closeWindow()
                    }
                })
            }
            if (drones.contains(5, true)) {
                interactions.put("Level 5", object : IInteractable.Interaction {
                    override fun interact() {
                        releaseDrone(5)
                        closeWindow()
                    }
                })
            }
        }

        return interactions
    }

    fun heal(powerLevel: Float) {
        health = MathUtils.clamp(health + powerLevel, 0f, maxHealth.toFloat()).toInt()
    }
}